### PR TITLE
WIP: Prom

### PR DESCRIPTION
**Pull request by kylebrandt**
```
$metric = "container_memory_working_set_bytes"
$tags   = "container_name"
$filter = '''container_name !~ "^$|pvc-.*$"'''
$agg    = "sum"
$step   = "1m"

$query = prom($metric, $tags, $filter, $agg, $step, "1h", "")
sort(avg($query), "desc") / 1000000
```

Too add:

- [x] `promrate`: like above but rate query, and with option for rate step
- [x] `promtags`: A list of keys, and the a map of keys to an array of values for the key, and a list of unique tag sets for give metric
- [x] `prommetrics`: A list of available metrics 
- [x] multiple datasource support (`promm` and `promratem`)
- [x] prom queries for query multiple datasources at once (`promm` and `promratem`)
- [x] Figure out what to do about subset results (e.g. you do `by (container,namespace)` and there are `{}` tagset results
- [x] Fix build (tests?)
- [x] Documentation for new funcs
- [x] Docs for prom system config (toml) 
- [x] Unit tests for template queries

My plan for the the prom queries for multiple data sources will be functions with name signatures but will accept a CSV in the prefix, e.g. `["default,staging,prod"]promm("container....)`. They will then tag their results with the label of the prom datasource (not sure what the tag key will be yet, don't know prom metrics well enough to know what is least likely to collide ... maybe "bosun_source"). With those results we should be able to use `aggr()` to aggregate across the multiple prom datasources if so desired.

Previous PR: https://github.com/bosun-monitor/bosun/pull/2143

### CODE REVIEW

**Comment by kylebrandt** [comment]
https://github.com/bosun-monitor/bosun/pull/2143/commits/e8f710a7103df0160929947d6100c96a7e6205b2

**Comment by kylebrandt** [comment]
@captncraig if using GH's view to view the diff, you want to expand expr/prom.go for the bulk of the code.

File: cmd/bosun/conf/system.go [hunk 1]
```diff
@@ -205,6 +209,24 @@ type InfluxConf struct {
...
+	URL string
+}
+
+// Valid returns if the configuration for the PromConf has required fields needed
+// to create a prometheus tsdb client
+func (pc PromConf) Valid() error {
+	if pc.URL == "" {
+		return fmt.Errorf("missing URL field")
+	}
+	_, err := api.NewClient(api.Config{Address: pc.URL})
```
[the remark this answers is not in the document]
Comment (kylebrandt) [reply cmd/bosun/conf/system.go]: Nope we are just making sure it is valid, a connection is not created here. If we erred on the connection not being made here Bosun would not start which isn't what we want. Will add comment.

File: docs/expressions.md [hunk 2]
```diff
@@ -489,6 +489,129 @@ and those numbers created into a series.
...
+
+promrate($metric, $groupByTags, $filter, $agg, $rateStep, $step, "1h", "")
+```
+
+The above example would generate a PromQL query of `sum(rate( container_memory_working_set_bytes { container_name !~ "pvc-.*$" }  [1m] )) by ( container_name,namespace )`. The time range and step are sent via HTTP query parameters.
+
+### promm(metric, groupByTags, filter, agType, stepDuration, startDuration, endDuration string) seriesSet
+{: .exprFunc}
+
+promm (Prometheus Multiple) is like the `prom` function, except that it queries multiple Prometheus TSDBs and combines the result into a single seriesSet. A tag key of `bosun_prefix` with the tag value set to the prefix is added to the results to ensure that series are unique in the result.
```
[the remark this answers is not in the document]
Comment (kylebrandt) [reply docs/expressions.md]: It needs to be something that wouldn't already be label or there could be a collision. Something like "cluster" seems like a label/key that might already be used in metrics. http://bosun.org/expressions#renamevariantset-string-seriessetnumberset can be used to rename the tag key after the query.

File: cmd/bosun/expr/prom.go [hunk 3]
```diff
@@ -0,0 +1,442 @@
...
+			Value: values,
+			Group: tags,
+		})
+	}
+	return r, nil
+}
+
+// promQueryTemplate is a template for PromQL time series queries. It supports
+// filtering and aggregation
+var promQueryTemplate = template.Must(template.New("promQueryTemplate").Parse(`
```
[the remark this answers is not in the document]
Comment (kylebrandt) [reply cmd/bosun/expr/prom.go]: @captncraig isn't it better to have the template parsed only once, or is there a better to way to do that?

**Comment by captncraig** [comment]
LGTM

**Comment by pdf** [comment]
Is there a reason Prometheus ended up being handled differently to all the other TSDBs, in that it does not support native query strings?

**Comment by captncraig** [comment]
Not completely different than other tsdbs, but I see what you mean. I believe a primary reason is that bosun needs to know from the query exactly what the resulting tag sets will look like in order to do its type checking things on expressions. It is difficult to know what the resulting tag "shape" of a prometheus query will be from just the query, and it can still give inconsistent rows.

If "raw" prometheus queries are a highly demanded feature,we'd need to parse and analyse the query deeper, or manually specify the tags it returns. But there's still features like time handling that bosun needs to know about outside of the raw query as well. There's a number of complicated aspects that make a query-builder interface much easier.

**Comment by kylebrandt** [comment]
From an implementation standpoint on Bosun's side this seemed easier. Bosun
wants to know the type of response, seriesSet or numberSet, as well as the
tag keys of the response.

From a user standpoint I'm not so sure yet. This makes it so you can't just
copy and paste queries, but on the other hand having long query strings in
bosun expressions can be awkward as well.

That being said this style of query generating functions does not forever
have to be the only type of Prometheus queries inside of Bosun.

**Comment by pdf** [comment]
Even ignoring the pain for users of deconstructing queries to fit these function sigs (which is not insignificant), it's particularly problematic because you can't do complex aggregations (eg - calculations across series), so Bosun can't used to alert on the same queries used for graphing.

I'm not sure I'm completely in love with it, but one option might be to allow variable interpolation in the query strings if you need some data to be externally available, and parsing the query is too hard, eg:

```
$query = "sum(irate(node_disk_reads_completed_total[%%stepDuration%%]) + irate(node_disk_writes_completed_total[%%stepDuration%%])) by (%%groupByTags%%)"
$groupByTags = "instance"
$stepDuration = "5m"
promql($query, $groupByTags, $stepDuration, "1h", "")
```

(`%%` delims as an example, not sure what delimiters are viable in all parsing contexts off the top of my head)

More thought probably required though.

**Comment by kylebrandt** [comment]
@captncraig / @pdf Actually this might not be that hard if we have diff funcs for series vs number set:

```
"github.com/prometheus/prometheus/promql"

parsedExpr, err := promql.ParseExpr(myRawExpr)
if err != nil {
  ...
}
fmt.Println(parsedExpr.Type(), parsedExpr.(*promql.AggregateExpr).Grouping)
```

Would check assertions etc, but the `Grouping` property gives us keys and I think we can get what we need from the `Type()` call.

**Pull request merged**